Support paramatized yum::versionlock

traylenator opened the pull request:
#### Pull Request (PR) description

Support paramatized yum::versionlock

Motivated by the change of format to the versionlock file with DNF systems
change `yum::versionlock` so it supports parameters `version`,
`release`, `epoch` and `arch`. If version is specified the namevar is now the package name.

So on a DNF based systemd (e.g CentOS 8) the following must be used.

```puppet
yum::versionlock{'bash':
  version => '3.4',
}
```

results in a version lock of on a dnf based system

```
bash-0:3.4-*
```

and on a yum based system of

```
0:bash-3.4-*
```

A fuller example with all parameters set.

```puppet
yum::versionlock{'bash':
  version  => '5.4',
  release  => 'alpha-rc3',
  epoch    => 8,
  arch     => 'noarch'
}
```

results in a version lock on a DNF based system of

```
bash-8:5.4-alpha-rc3.noarch
```

The same mechanism on YUM based systems (e.g CentOS 7) can be used.
To specify this new machanism at least a `version` should be set.

```puppet
yum::versionlock{'bash':
  version  => '5.4',
  release  => 'alpha-rc3',
  epoch    => 8,
  arch     => 'noarch'
  future   => true,
}
```

would results in

```
8:bash-5.4-alpha-rc3.noarch
```

as used by yum. If `version` is left default `undef`  then the old mechanism is used so this
is backwards compatable. e.g

```puppet
yum::versionlock{"0:bash-3.4-7.el8.x86_64"}
```
is still valid



#### This Pull Request (PR) fixes the following issues
Fixes #150

<!--
Thank you for contributing to this project!

- This project has a Contributor Code of Conduct: https://voxpupuli.org/coc/
- Please check that here is no existing issue or PR that addresses your problem.
- Our vulnerabilities reporting process is at https://voxpupuli.org/security/

-->
<!--
Replace this comment with a description of your pull request.
-->

<!--
Replace this comment with the list of issues or n/a.
Use format:
Fixes #123
Fixes #124
-->

traylenator commented:
Currently if 

```puppet
yum::versionlock{'0:bash-4.1.2-9.el6_2.x86_64':}
```
is used this compiles on CentOS 8 but is almost certainly not what was wanted.
Can we make the assumption that package names never contain `:` . It does seem to be case
for Fedora.

traylenator commented:
I tried to enable CentOS 8 accept tests but the framework hangs early.

vchepkov commented:
IMHO, `future` attribute is unnecessary. resource should make decision what format to use based on package provider. 'dnf' provider should use new format

vchepkov commented on manifests/versionlock.pp:
```diff
@@ -1,39 +1,103 @@
+  Yum::RpmVersion                      $version     = '*',
+  Yum::RpmRelease                      $release     = '*',
+  Variant[Yum::RpmEpoch,Enum['*']]     $epoch       = 0,
+  Optional[Yum::RpmArch]               $arch        = undef,
```
This also can be `*`

traylenator replied:
Is `*` different to `undef` ?  

Are these the same in reality or different?

```
name-0:acd-def.*
```
is different to

```
name-0:abc-def
```

fine to add however.

vchepkov replied:
But Yum::RpmArch wouldn't allow *, right?

traylenator replied:
I was questioning if needed since undef and `*` are equivalent but it makes sense to add in case they are not.

traylenator commented:
>  IMHO, future attribute is unnecessary. resource should make decision what format to use based on package provider. 'dnf' provider should use new format

It is in the sense you can ignore it. 

On 8 you get the new behaviour.
On 7 you get the old behaviour.

You can choose to get the new behaviour on 7 as well.

$package_provider is is totally used to decide a yum or dnf style format.

vchepkov commented on README.md:
````diff
@@ -309,7 +309,9 @@ yum::plugin { 'versionlock':
 ```
 
 ### Lock a package with the *versionlock* plugin
+The `versionlock` type changed between CentOS 8 and CentOS8.
````
You probably meant something different here

vchepkov commented on REFERENCE.md:
```diff
@@ -546,26 +551,55 @@ Default value: `undef`
+yum::versionlock { 'bash':
+  ensure => present,
+  version => '4.1.2',
+  release => '9.el6',
```
one probably wouldn't lock el6 version on CentOS 8?

traylenator replied:
This and other el6, 7 , 8 examples made consistent with example package names to contain el6, 7 or 8.

vchepkov commented:
Not sure why you would ever set 'future' on EL7, since it will render configuration invalid.

vchepkov commented on README.md:
```diff
@@ -424,4 +442,4 @@ This module was donated by CERIT Scientific Cloud, <support@cerit-sc.cz> to Vox
 [4]: https://tickets.puppetlabs.com/browse/PUP-3323
-[5]: ttps://puppet.com/docs/bolt/latest/bolt.html
\ No newline at end of file
+[5]: ttps://puppet.com/docs/bolt/latest/bolt.html
```
https?

traylenator commented:
Sorry for delay - long weekend here.

> Ah, I think I got what you are trying to do. IMHO, instead of adding 'future' attribute, why not deduct it from the fact that any of the new attributes are set instead of undef?

My first iteration of this I set arch, release version to undef as default and then used one of them being set to infer we were in modern mode. I changed as I prefer the `*` defaults as its more useful in the future - a better final interface.

Can I draw your attention to the `:` in package name comment [above](https://github.com/voxpupuli/puppet-yum/pull/169#issuecomment-624924726) as I think that is the best way to drop the need or desire for the a `future` parameter on 7.

> Ah, I think I got what you are trying to do.

Just to confirm I my motivation was to avoid maintaining

```puppet
if versioncmp($facts['os']['release']['version'],'7') <= 0 {
   yum::versionlock('0:bash-1.2.3.el6':}
} else {
  yum::versionlock('bash':
    version => '1.2.3',
    release  => 'el6',
}
```

Instead folk can just do for 6 , 7 and 8

```puppet
yum::versionlock('bash':
    version => '1.2.3',
    release  => 'el6',
    future    => true,
}
```

with this patch as it stands people have the choice of doing either above.

traylenator commented:
In principle this patch supports someone installing dnf on CentOS 7 as well which is now available. Can't comment on the rest of the module mind you.

vchepkov commented:
I understand. IMHO, you can set default  value for `version` to `undef` and use that as indicator of the _future_ format instead.

vchepkov commented:
> In principle this patch supports someone installing dnf on CentOS 7 as well which is now available. Can't comment on the rest of the module mind you.

Oh, haven't realized that. It's not in any major repositories. In this case maybe parameter should be  `Enum['dnf','yum'] $format = 'yum'` ?

traylenator commented:
> I understand. IMHO, you can set default value for version to undef and use that as indicator of the future format instead.

Seems sensibe. Indeed I can't think of use case where you would set release to something and leave version as `*` and even if you did want that you could change version to `*`.

I'll make those changes.


> Oh, haven't realized that. It's not in any major repositories. In this case maybe parameter should be Enum['dnf','yum'] $format = 'yum' ?

I think we would have to assume that  `package_provider` fact would be changed to dnf on dnf installed on 7 machine.  Certainly assume it until someone actually tries it. Not something I will do until CentOS 6 is dead and out the window.

vchepkov commented:
If this new parameter is just to accommodate EL7 users with dnf installed, then disregard my last comment. 

pacakge_provider comes from stdlib module, but it just uses logic from the puppet itself

https://github.com/puppetlabs/puppet/blob/master/lib/puppet/provider/package/dnf.rb#L31-L34

According to it, package_provider will be set to yum on EL7

vchepkov commented:
Another option is to drop the old format and make this incompatible change release.
The old code is bad, since you are allowed to 'lock' multiple versions of the same package without ever noticing.

vchepkov commented:
You should probably remove that future attribute, since you don't use it anymore.
Other than that looks good!

traylenator commented:
Thanks future parameter removed.

I'll add some strings doc to the new types, rebase , rewrite the description and hopefully there.

ekohl commented on REFERENCE.md:
```diff
@@ -660,6 +726,62 @@ The hash on which to operate
+### Yum::RpmEpoch
+
+valid epoch of rpm, it is always and integer greater than 0
+example 0
+example 1
```
This renders in an odd way

ekohl commented on REFERENCE.md:
````diff
@@ -660,6 +726,62 @@ The hash on which to operate
+##### openssh
+
+```puppet
+
````
Missing?

traylenator replied:
Bit lost about this one. Maybe a rendering one.

ekohl commented on REFERENCE.md:
````diff
@@ -660,6 +726,62 @@ The hash on which to operate
+##### python35-foobar
+
+```puppet
+
````
Missing?

ekohl commented on manifests/versionlock.pp:
```diff
@@ -1,39 +1,106 @@
+  Enum['present', 'absent', 'exclude']      $ensure  = 'present',
+  Optional[Yum::RpmVersion]                 $version = undef,
+  Yum::RpmRelease                           $release = '*',
+  Variant[Yum::RpmEpoch,Enum['*']]          $epoch   = 0,
+  Optional[Variant[Yum::RpmArch,Enum['*']]] $arch    = undef,
```
Personal preference is to have a space after the comma so visually separate it.

ekohl commented on types/rpmarch.pp:
```diff
@@ -0,0 +1,5 @@
+# Possible rpm architectures
+# See https://github.com/rpm-software-management/rpm/blob/master/rpmrc.in
```
```suggestion
# @summary Possible rpm architectures
# @see https://github.com/rpm-software-management/rpm/blob/master/rpmrc.in
```

ekohl commented on types/rpmarch.pp:
```diff
@@ -0,0 +1,5 @@
+# Possible rpm architectures
+# See https://github.com/rpm-software-management/rpm/blob/master/rpmrc.in
+# example noarch
+# example x86_64
```
I'm not sure this adds anything for a trivial enum

ekohl commented on types/rpmepoch.pp:
```diff
@@ -0,0 +1,4 @@
+# valid epoch of rpm, it is always and integer greater than 0
+# example 0
+# example 1
```
I don't think these examples add anything

ekohl commented on types/rpmname.pp:
```diff
@@ -0,0 +1,4 @@
+# Valid RPM names can be almost anything
+# Example openssh
+# Example python35-foobar
```
Looks like this it interpreted and causes the odd rendering. Not sure what the best solution is.

traylenator commented:
@ekohl Indeed the regexs on version, release , ... absolutely needed a `^` and `$`.

* Have added type_aliases tests now.
* Have improved the rendering of type string data now I hope but https://github.com/voxpupuli/puppet-yum/pull/171 is needed first to stop current rendering failure to really confirm.

vchepkov commented:
- puppetlabs/stdlib just recently went through replacing all regex with ^ and $ to \A and \z instead
- epoch can't be '*', in fact you added test in line 8 in yum_rpmepoch_spec.rb  that it can't be star, but Variant allows it
- line with assert_type(Yum::RpmVersion is redundant, this is already done by type validation
- I am not sure where the symbols that are allowed  by type Yum::RpmName are coming from.
the referenced spec file doesn't mention it, but according https://fedoraproject.org/wiki/Packaging:Naming?rd=Packaging:NamingGuidelines#Common_Character_Set_for_Package_Naming  only few symbols are allowed in the name

ekohl commented:
> puppetlabs/stdlib just recently went through replacing all regex with ^ and $ to \A and \z instead

Just to help me understand: what's the reason for this?

vchepkov commented:
> > puppetlabs/stdlib just recently went through replacing all regex with ^ and $ to \A and \z instead
> 
> Just to help me understand: what's the reason for this?

It was just an observation. I presume it's to avoid allowing a multiline string ?

traylenator commented:
>  puppetlabs/stdlib just recently went through replacing all regex with ^ and $ to \A and \z instead

updated.

> epoch can't be '*', in fact you added test in line 8 in yum_rpmepoch_spec.rb that it can't be star, but Variant allows it
line with assert_type(Yum::RpmVersion is redundant, this is already done by argument type validation

That makes sense to me. RPM's do not support arch of `*` but versionlock files do
support a `*` for arch.

> I am not sure where the symbols that are allowed by type Yum::RpmName are coming from.
the referenced spec file doesn't mention it, but according https://docs.fedoraproject.org/en-US/packaging-guidelines/Naming/ only few symbols are allowed in the name

I think they were lifted from the previous versionlock type.

Above link is a guideline rather than enforced. If I try and make the RPM with a crazy name.

```
Name: p.u_p+p%%e{t}ag-ent
```
I do indeed get an RPM 

```
/home/steve/rpmbuild/SRPMS/p.u_p+p%e{t}ag-ent-5.5.19.5.g340c40ba-1.fc32.src.rpm
```

:-)

The `*` does cause an exception so I'll remove the * as an option from name now.

traylenator commented:
> line with assert_type(Yum::RpmVersion is redundant, this is already done by argument type validation

agreed.

traylenator commented:
l> line with assert_type(Yum::RpmVersion is redundant, this is already done by argument type validation

> agreed.

It is needed afterall. The argument type allows undef for a yum based machine but on a
dnf based machine version must be set.

@vchepkov I believe I have addressed your comments.

vchepkov commented:
> It is needed afterall. The argument type allows undef for a yum based machine but on a
> dnf based machine version must be set.
> 

IMHO, you don't need it, but it's not important, it's just a matter of preference.
In your case value of undef is unacceptable, the type assertion is not necessary.
also, I wouldn't ascertain $version twice. pseudo code:
```
  if $version {
   # Variable is defined, must be a new style and pattern validated by type
    case $facts['package_provider'] {
       'yum': { yum pattern }
       'dnf': { dnf pattern }
       default: { fail not supported yet}
    }
  } else {
  # old style, assert name type
 }
```

vchepkov commented:
> That makes sense to me. RPM's do not support arch of `*` but versionlock files do
> support a `*` for arch.

I was talking about `epoch`, which can't be '*'
```
[root@puppet ~]# cat /etc/yum/pluginconf.d/versionlock.list 
puppet-agent-*:6.14.0-1.el8.*
[root@puppet ~]# yum update
Last metadata expiration check: 0:03:32 ago on Tue 12 May 2020 12:49:55 PM UTC.
Versionlock plugin: could not parse pattern: puppet-agent-*:6.14.0-1.el8.*

```

traylenator commented:
> > That makes sense to me. RPM's do not support arch of `*` but versionlock files do
> > support a `*` for arch.
> 
> I was talking about `epoch`, which can't be '*'
> 
> ```
> [root@puppet ~]# cat /etc/yum/pluginconf.d/versionlock.list 
> puppet-agent-*:6.14.0-1.el8.*
> [root@puppet ~]# yum update
> Last metadata expiration check: 0:03:32 ago on Tue 12 May 2020 12:49:55 PM UTC.
> Versionlock plugin: could not parse pattern: puppet-agent-*:6.14.0-1.el8.*
> ```

traylenator commented:
> I was talking about epoch, which can't be '*'

epoch is now just an `Integer[0]`

traylenator commented:
Have squashed commits , will merge Monday if no more comment.

vchepkov commented on REFERENCE.md:
```diff
@@ -568,6 +601,39 @@ Specifies if versionlock should be `present`, `absent` or `exclude`.
+
+##### `epoch`
+
+Data type: `Variant[Yum::RpmEpoch, Enum['*']]`
```
forgot to update this line

ekohl commented on spec/defines/versionlock_spec.rb:
```diff
@@ -1,82 +1,184 @@
+      context 'with version set namevar must be just a package name' do
+        let(:params) { { version: '4.3' } }
+
+        it { is_expected.to raise_error(Puppet::PreformattedError, %r{Package name must be formatted as Yum::RpmName, not 'String'}) }
```
I've started to like this:
```suggestion
        it { is_expected.to compile.and_raise_error(%r{Package name must be formatted as Yum::RpmName, not 'String'}) }
```

There's a few instances and I won't repeat it everywhere.

traylenator replied:
@ekohl I completly missed these comments before I pressed merge. 

My apologies to excited to merge :-)

I will of course work your comments into a fresh merge request.
